New option to generate reports in Markdown format `--markdown`

guillermocalvo opened the pull request:
Solves #404

> [!NOTE]
> An example Markdown report generated for demonstration purposes can be found [here](https://github.com/guillermocalvo/japicmp/actions/runs/10148680458#summary-28061943074).

@siom79 Sorry for the big PR 😅 I tried to make this new functionality as customizable as possible. Please feel free to edit it or suggest any changes you think should be done before it can be merged.

scordio commented:
Very cool work @guillermocalvo, thanks for taking this up!

siom79 commented on src/site/markdown/Examples.md:
````diff
@@ -30,6 +30,11 @@ are marked with a trailing star (*).
 ...
 ```
 
+At your preference, japicmp can generate a [Markdown](https://www.markdownguide.org/) report.
````
Can you move this section below the HTML section, i.e. between HTML and XML? :smirk:

guillermocalvo replied:
Sure https://github.com/siom79/japicmp/pull/405/commits/98fe35bcd0d6c509f8dccf5231631f91d784844e

siom79 commented on japicmp/src/main/java/japicmp/output/markdown/MarkdownOutputGenerator.java:
```diff
@@ -0,0 +1,606 @@
+
+	@Override
+	public String generate() {
+		return toString();
```
Why override `toString()` here?

guillermocalvo replied:
Got it. I simplified it here: https://github.com/siom79/japicmp/pull/405/commits/e00fd6afa0cb73dd8a4d3034c058c2196eb96cc6

siom79 commented on japicmp/src/main/java/japicmp/util/JApiClassFileFormatVersionHelper.java:
```diff
@@ -0,0 +1,71 @@
+		} else if (major == 49) {
+			return "JDK 5.0";
+		} else if (major == 50) {
+			return "JDK 6";
```
What about `return "JDK " + (major - 44)`?

guillermocalvo replied:
Sounds good. I just refactored it here: https://github.com/siom79/japicmp/pull/405/commits/c4659f1b9f89a95d297a7b3df752cb5ca9ebd03b

siom79 reviewed (changes requested):
Excellent work! Thanks very much.

guillermocalvo commented on japicmp/src/main/java/japicmp/cli/JApiCli.java:
```diff
@@ -46,6 +47,12 @@ private void generateOutput(Options options, List<JApiClass> jApiClasses, JarArc
+			MarkdownOutputGenerator markdownOutputGenerator = new MarkdownOutputGenerator(options, jApiClasses);
+			String output = markdownOutputGenerator.generate();
+			System.out.println(output);
+			return;
```
You're right. It's better that way. https://github.com/siom79/japicmp/pull/405/commits/1c73c30a9474c077807cd7e0ccc63b0017ec9692

guillermocalvo commented:
> Excellent work! Thanks very much.

Thanks for your review @siom79! I have implemented your suggestions.

Incidentally, [the JDK 8 build failed](https://github.com/siom79/japicmp/actions/runs/10476545924/job/29015660610) because the JVM was confused about some Java types 🤔 It's weird, because [it was working perfectly fine two weeks ago](https://github.com/siom79/japicmp/actions/runs/10235693979/job/28316835660), so I think the CI worker must be running a different version now. In any case, [I extracted an expression to a constant](https://github.com/siom79/japicmp/pull/405/commits/554a6c6c046ccf35c60daec92ca16bcca8ff48ba) to make sure the compiler understands what the types in play are.

siom79 commented on japicmp/src/main/java/japicmp/cli/JApiCli.java:
```diff
@@ -69,6 +70,12 @@ private void generateOutput(Options options, List<JApiClass> jApiClasses, JarArc
+			MarkdownOutputGenerator markdownOutputGenerator = new MarkdownOutputGenerator(options, jApiClasses);
+			String output = markdownOutputGenerator.generate();
+			System.out.println(output);
+			return;
```
I would remove the `return` and make an else case here, as otherwise the `IncompatibleErrorOutput` is not executed.

siom79 commented:
Thanks @guillermocalvo for implementing the changes. I have resolved the two last remarks myself and merged the result to main.

Btw: Are you interested to become a collaborator on this repository?

guillermocalvo commented:
> Btw: Are you interested to become a collaborator on this repository?

@siom79 Sounds great! Is there anything specific I can help with?

siom79 commented:
@guillermocalvo I have sent you an invitation. Currently there is nothing specific, but as contributor you can merge PRs and do other helpful things. But please be cautious with the new power. :smirk:

guillermocalvo commented:
@siom79 Thanks! And don't worry, I know that "with great power comes great responsibility". 😉
